### PR TITLE
 [mlir][sparse] first proof-of-concept non-permutation rewriter

### PR DESCRIPTION
**Pull request by aartbik**
Rather than extending sparsifier codegen with higher order non-permutations, we follow the path of rewriting linalg generic ops into higher order operations. That way, code generation will simply work out of the box. This is a very first proof-of-concept rewriting of that idea.

### CODE REVIEW

File: mlir/lib/Dialect/SparseTensor/Transforms/SparseReinterpretMap.cpp [hunk 1]
```diff
@@ -18,10 +20,135 @@ using namespace mlir::sparse_tensor;
...
+
+// Generates a "de"mapping reinterpretation of the map.
+static Value genDemap(OpBuilder &builder, SparseTensorEncodingAttr enc,
+                      Value val) {
+  unsigned lvlRank = enc.getLvlTypes().size();
+  AffineMap idMap =
+      AffineMap::getMultiDimIdentityMap(lvlRank, builder.getContext());
+  auto newEnc = SparseTensorEncodingAttr::get(
+      builder.getContext(), enc.getLvlTypes(), idMap, idMap, enc.getPosWidth(),
+      enc.getCrdWidth());
```
Comment (PeimingLiu) [inline mlir/lib/Dialect/SparseTensor/Transforms/SparseReinterpretMap.cpp:49]: I find that there is a convenient method for this: `enc.withoutDimToLvl`

File: mlir/lib/Dialect/SparseTensor/Transforms/SparseReinterpretMap.cpp [hunk 2]
```diff
@@ -18,10 +20,135 @@ using namespace mlir::sparse_tensor;
...
-// TODO:
-//   (1) insert the zero-cost sparse_tensor.reinterpret_map ops
-//   (2) rewrite linalg.generic ops traits on level crds
-//   (3) compute topsort, and resolve cyles with sparse_tensor.convert ops
+//===----------------------------------------------------------------------===//
+// Helper methods.
+//===----------------------------------------------------------------------===//
+
+// Translates a "simple" map according to an identify lvl-map.
+static AffineMap translateMap(OpBuilder &builder, SparseTensorType stt,
```
Comment (PeimingLiu) [inline mlir/lib/Dialect/SparseTensor/Transforms/SparseReinterpretMap.cpp:28]: nit: I don't think you need `static` here as it is already inside an anonymous namespace

File: mlir/lib/Dialect/SparseTensor/Transforms/SparseReinterpretMap.cpp (hunk 1 above)
Comment (aartbik) [reply mlir/lib/Dialect/SparseTensor/Transforms/SparseReinterpretMap.cpp:49]: Yes, I noticed your other helpers for this too. But let me make this change at least.

File: mlir/lib/Dialect/SparseTensor/Transforms/SparseReinterpretMap.cpp (hunk 2 above)
Comment (aartbik) [reply mlir/lib/Dialect/SparseTensor/Transforms/SparseReinterpretMap.cpp:28]: Oh yeah, I am old-fashioned in the sense that "static" to me immediately flags the file-static property more so than the namespace context ;-) But as a result, I think it has become more or less style in this directory ;-)

Comment (PeimingLiu) [reply mlir/lib/Dialect/SparseTensor/Transforms/SparseReinterpretMap.cpp:28]: I triggered clang-tidy warning on my editor though. (but won't fail clang-format)

File: mlir/test/Dialect/SparseTensor/sparse_reinterpret_map.mlir [hunk 3]
```diff
@@ -8,3 +8,50 @@
 func.func @sparse_nop(%arg0: tensor<?xf64, #SparseVector>) -> tensor<?xf64, #SparseVector> {
   return %arg0 : tensor<?xf64, #SparseVector>
 }
```
Comment (PeimingLiu) [inline mlir/test/Dialect/SparseTensor/sparse_reinterpret_map.mlir:10]: How about deleting it now?

**Review by PeimingLiu — APPROVED** [review]

File: mlir/test/Dialect/SparseTensor/sparse_reinterpret_map.mlir (hunk 3 above)
Comment (aartbik) [reply mlir/test/Dialect/SparseTensor/sparse_reinterpret_map.mlir:10]: Ok, next one (when we have > 1 ;-)

**Pull request merged**